File: 2025/B/answers/java/Solution.java
import java.util.*;

public class Solution {
    static final int MOD = 998244353;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextLong()) return;
        long N = sc.nextLong();
        int K = sc.nextInt();

        int[] dp = new int[K + 1];
        for (int q = 1; q <= K; q++) dp[q] = 1;

        int limit = (int) Math.min(N, (long) K);

        for (int i = 2; i <= limit; i++) {
            int[] newDpDiff = new int[K + 2];
            for (int qOld = 1; qOld <= K; qOld++) {
                if (dp[qOld] == 0) continue;

                int low = qOld - (qOld - 1) / i;
                int high = qOld + (K - qOld) / i;

                if (low <= K) {
                    newDpDiff[low] = (newDpDiff[low] + dp[qOld]) % MOD;
                    if (high + 1 <= K) {
                        newDpDiff[high + 1] = (newDpDiff[high + 1] - dp[qOld] + MOD) % MOD;
                    }
                }
            }
            int curr = 0;
            for (int q = 1; q <= K; q++) {
                curr = (curr + newDpDiff[q]) % MOD;
                dp[q] = curr;
            }
        }

        long total = 0;
        for (int q = 1; q <= K; q++) {
            total = (total + dp[q]) % MOD;
        }
        System.out.println(total);
    }
}

File: 2025/B/answers/kt/Solution.kt
import java.util.*

val MOD = 998244353

fun main() {
    val sc = Scanner(System.`in`)
    if (!sc.hasNextLong()) return
    val n = sc.nextLong()
    val k = sc.nextInt()

    var dp = IntArray(k + 1) { 1 }
    dp[0] = 0

    val limit = Math.min(n, k.toLong()).toInt()

    for (i in 2..limit) {
        val newDpDiff = IntArray(k + 2)
        for (qOld in 1..k) {
            if (dp[qOld] == 0) continue

            val low = qOld - (qOld - 1) / i
            val high = qOld + (k - qOld) / i

            if (low <= k) {
                newDpDiff[low] = (newDpDiff[low] + dp[qOld]) % MOD
                if (high + 1 <= k) {
                    newDpDiff[high + 1] = (newDpDiff[high + 1] - dp[qOld] + MOD) % MOD
                }
            }
        }
        var curr = 0
        for (q in 1..k) {
            curr = (curr + newDpDiff[q]) % MOD
            dp[q] = curr
        }
    }

    var total = 0L
    for (q in 1..k) {
        total = (total + dp[q]) % MOD
    }
    println(total)
}
